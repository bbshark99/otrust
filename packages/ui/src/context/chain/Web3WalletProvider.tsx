import React, { useMemo } from 'react';
import { useWebWalletBackend, WebWalletBackendProvider } from '@onomy/wallet-backend-web';
import { BondingCurveProvider, OnomyHubProvider } from '@onomy/react-hub';

import Landing from 'pages/Landing';
import {
  KEPLR_CONFIG,
  REACT_APP_BONDING_NOM_CONTRACT_ADDRESS,
  REACT_APP_ETH_WS_URL,
  REACT_APP_GRAVITY_CONTRACT_ADDRESS,
  REACT_APP_ONOMY_RPC_URL,
  REACT_APP_WNOM_CONTRACT_ADDRESS,
} from 'constants/env';

function Web3WalletChild({ children }: { children: JSX.Element | JSX.Element[] }) {
  const backend = useWebWalletBackend();
  const chainId = useMemo(() => {
    return KEPLR_CONFIG.chainId;
    const pieces = KEPLR_CONFIG.chainId.split('-');
    pieces.pop();
    return pieces.join('-');
  }, []);

  if (!backend) return null;

  return (
    <OnomyHubProvider
      backend={backend}
      onomyChainId={chainId}
      onomyChainInfo={KEPLR_CONFIG}
      onomyRpcUrl={REACT_APP_ONOMY_RPC_URL}
      ethWsUrl={REACT_APP_ETH_WS_URL}
      nomContractAddress={REACT_APP_WNOM_CONTRACT_ADDRESS}
      bondContractAddress={REACT_APP_BONDING_NOM_CONTRACT_ADDRESS}
      gravityContractAddress={REACT_APP_GRAVITY_CONTRACT_ADDRESS}
    >
      <BondingCurveProvider>{children}</BondingCurveProvider>
    </OnomyHubProvider>
  );
}

export function Web3WalletProvider({ children }: { children: JSX.Element | JSX.Element[] }) {
  return (
    <WebWalletBackendProvider Landing={Landing}>
      <Web3WalletChild>{children}</Web3WalletChild>
    </WebWalletBackendProvider>
  );
}
