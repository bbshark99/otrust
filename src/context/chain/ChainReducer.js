import { isBigNumber } from 'bignumber.js';

function reducerCallback(state, key, value, update) {
  switch (state[key]) {
    case value:
      break;
    default:
      switch (true) {
        case isBigNumber(value):
          return {
            [key]: value,
            ...update,
          };
        default:
          throw new Error();
      }
  }
}

export function reducer(state, action) {
  let update;
  switch (action.type) {
    case 'updateAll':
      // eslint-disable-next-line no-restricted-syntax
      for (const [key, value] of action.value.entries()) {
        if (state[key]) {
          try {
            switch (key) {
              case 'currentETHPrice':
                update = reducerCallback(state[key], key, value, update);
                break;
              case 'currentNOMPrice':
                update = reducerCallback(state[key], key, value, update);
                break;
              case 'NOMallowance':
                update = reducerCallback(state[key], key, value, update);
                // console.log('Update NOM Allowance: ', update);
                break;
              case 'strongBalance':
                update = reducerCallback(state[key], key, value, update);
                break;
              case 'supplyNOM':
                update = reducerCallback(state[key], key, value, update);
                break;
              case 'weakBalance':
                update = reducerCallback(state[key], key, value, update);
                break;
              case 'blockNumber':
                update = reducerCallback(state[key], key, value, update);
                break;
              default:
                throw new Error('Unexpected default case');
            }
          } catch (e) {
            // eslint-disable-next-line no-console
            console.error(e);
          }
        }
      }
      // console.log('ChainReducer Update: ', update);
      if (update) {
        return {
          ...update,
        };
      }
      break;
    default:
      throw new Error();
  }
}
