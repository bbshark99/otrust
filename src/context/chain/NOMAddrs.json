{
  "NOMERC20": "0xe7c0fd1f0A3f600C1799CD8d335D31efBE90592C",
  "BondingNOM": "0x52A4A4FBB36A2f6A83dBf7De9425C1ec5FF528Da",
  "Gravity": "0x8497452388c710Ebecd09adBF018C41f1Fe69876"
}
