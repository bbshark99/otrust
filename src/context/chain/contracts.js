import { Contract } from '@ethersproject/contracts';

import {
  BONDING_NOM_CONTRACT_ADDRESS,
  GRAVITY_CONTRACT_ADDRESS,
  WNOM_CONTRACT_ADDRESS,
} from 'constants/env';

const NomContractJSON = require('./ERC20NOM.json');
const BondingContractJSON = require('./BondingNOM.json');
const UniswapContractJSON = require('./UniSwap.json');
const GravityContractJSON = require('./Gravity.json');

const uniswapUsdcAddress = '0xb4e16d0168e52d35cacd2c6185b44281ec28c9dc';
/**
 * NOM ERC20 Contract instance
 */
export function NOMCont(library) {
  const ABI = NomContractJSON.abi;
  return new Contract(WNOM_CONTRACT_ADDRESS, ABI, library?.getSigner());
}

/**
 * Haven Contract instance
 */
export function BondingCont(library) {
  const ABI = BondingContractJSON.abi;
  return new Contract(BONDING_NOM_CONTRACT_ADDRESS, ABI, library?.getSigner());
}

/**
 * UniSwap Contract instance
 */
export function UniSwapCont(library) {
  const ABI = UniswapContractJSON;
  return new Contract(uniswapUsdcAddress, ABI, library?.getSigner());
}

export function GravityCont(library) {
  const ABI = GravityContractJSON.abi;
  return new Contract(GRAVITY_CONTRACT_ADDRESS, ABI, library.getSigner());
}
