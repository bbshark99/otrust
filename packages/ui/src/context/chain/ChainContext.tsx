import React from 'react';

import { Web3WalletProvider } from './Web3WalletProvider';
// import { JsWalletProvider } from './JsWalletProvider';

export function AppProvider({ children }: { children: JSX.Element | JSX.Element[] }) {
  return <Web3WalletProvider>{children}</Web3WalletProvider>;
}

export default AppProvider;
