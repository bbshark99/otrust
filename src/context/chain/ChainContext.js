import React, { useReducer, useEffect, createContext, useContext } from 'react'
import { useWeb3React } from "@web3-react/core"
import ApolloClient, { InMemoryCache } from 'apollo-boost'
import { ApolloProvider } from '@apollo/client'
import { BondingCont, NOMCont, contAddrs } from 'context/chain/contracts'
import { BigNumber } from 'bignumber.js'
import { reducer } from 'context/chain/ChainReducer'

export const ChainContext = createContext()
export const useChain = () => useContext(ChainContext)

export const UpdateChainContext = createContext()
export const useUpdateChain = () => useContext(UpdateChainContext)

function ChainProvider({ theme, children }) {
    
    const { account, library } = useWeb3React()
    const bondContract = BondingCont(library)
    const NOMContract = NOMCont(library)
    const [state, dispatch] = useReducer(reducer, { 
        blockNumber: new BigNumber(0),
        currentETHPrice: new BigNumber(0),
        currentNOMPrice: new BigNumber(0),
        gasOptions: [
            {
            id: 0,
            text: "0 (Standard)",
            },
            {
            id: 1,
            text: "0 (Fast)",
            },
            {
            id: 2,
            text: "0 (Instant)",
            },
        ],
        NOMallowance: new BigNumber(0),
        strongBalance: new BigNumber(0),
        supplyNOM: new BigNumber(0),
        weakBalance: new BigNumber(0)
    })
    
    if (!process.env.REACT_APP_GRAPHQL_ENDPOINT) {
        throw new Error('REACT_APP_GRAPHQL_ENDPOINT environment variable not defined')
    }

    const client = new ApolloClient({
        uri: process.env.REACT_APP_GRAPHQL_ENDPOINT,
        cache: new InMemoryCache(),
    })

    const getGasOptions = async () => {
        const prices = await fetch('https://www.gasnow.org/api/v3/gas/price?utm_source=onomy');
            const result = await prices.json();
        
        let gasOptions = [
            {
                id: 0,
                text: "0 (Standard)",
                gas: new BigNumber(0)
            },
            {
                id: 1,
                text: "0 (Fast)",
                gas: new BigNumber(0)
            },
            {
                id: 2,
                text: "0 (Instant)",
                gas: new BigNumber(0)
            }
        ]
                    
        gasOptions[0].text = result.data.standard / 1e9 + " (Standard)";
        gasOptions[1].text = result.data.fast / 1e9 + " (Fast)";
        gasOptions[2].text = result.data.rapid / 1e9 + " (Instant)";
        gasOptions[0].gas = new BigNumber((result.data.standard / 1e9).toString())
        gasOptions[1].gas = new BigNumber((result.data.fast / 1e9).toString())
        gasOptions[2].gas = new BigNumber((result.data.rapid / 1e9).toString())

        return gasOptions
    }

    useEffect(() => {
        // listen for changes on an Ethereum address
        library.on('block', async (number) => {
            if (state.blocknumber !== number) {
                try {
                    await Promise.all(
                        [
                            // Current ETH Price & Current NOM Price
                            bondContract.buyQuoteETH((10**18).toString()),
                            // NOM Allowance
                            NOMContract.allowance(account, contAddrs.BondingNOM),
                            // Strong Balance
                            library.getBalance(account),
                            // Supply NOM
                            bondContract.getSupplyNOM(),
                            // Weak Balance (May need to move these to Exchange)
                            NOMContract.balanceOf(account),
                            number,
                            getGasOptions()
                            // UniSwap Pricing
                            // UniSwapCont.getReserves(),
                        ]
                    ).then(values => {
                        let update = new Map()
                        for (let i = 0; i < values.length; i++) {
                            switch (i) {
                                case 0: 
                                    update = update.set(
                                        'currentETHPrice', 
                                        new BigNumber(values[0].toString())
                                    )

                                    update = update.set(
                                        'currentNOMPrice', 
                                        (new BigNumber(1)).div(new BigNumber(values[0].toString()))
                                    )
                                    break

                                case 1: 
                                    update = update.set(
                                        'NOMallowance', 
                                        new BigNumber(values[1].toString())
                                    )
                                    break

                                case 2: 
                                    update = update.set(
                                        'strongBalance',
                                        new BigNumber(values[2].toString())
                                    )
                                    break

                                case 3: 
                                    update = update.set(
                                        'supplyNOM', 
                                        new BigNumber(values[3].toString())
                                    )
                                    break

                                case 4: 
                                    update = update.set(
                                        'weakBalance', 
                                        new BigNumber(values[4].toString())
                                    )
                                    break

                                case 5:
                                    update = update.set(
                                        'blockNumber', 
                                        new BigNumber(number.toString())
                                    )
                                    break
                                case 6:
                                    update = update.set(
                                        'gasOptions', 
                                        values[6]
                                    )
                                    break
                                default: break
                            }
                        }
                        dispatch({type: 'updateAll', value: update})
                    })
                } catch {
                    console.log("Failed Chain Promise")
                }
            }
        })
        // remove listener when the component is unmounted
        return () => {
          library.removeAllListeners('block')
        }
        // trigger the effect only on component mount
    })
    

    const contextValue = {
        ...state,
        theme
    }

    return (
        <ApolloProvider client={client}>
            <UpdateChainContext.Provider value={dispatch}>
                <ChainContext.Provider value={contextValue} >
                    {children}
                </ChainContext.Provider>
            </UpdateChainContext.Provider>
        </ApolloProvider>
    )
}

export default ChainProvider
