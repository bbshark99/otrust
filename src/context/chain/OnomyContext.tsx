import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { BigNumber } from 'bignumber.js';

import { useStateRef } from 'hooks/useStateRef';
import {
  KEPLR_CONFIG,
  REACT_APP_ONOMY_REST_URL,
  REACT_APP_ONOMY_WS_URL,
  DENOM,
} from 'constants/env';
// eslint-disable-next-line import/no-cycle
import { ChainContext } from './ChainContext';
import { format18 } from 'utils/math';
import { OnomyClient } from 'OnomyClient';
import { useKeplr } from 'hooks/useKeplr';

// This is lame, but can't find a way to subscribe to cosmos events
const POLLING_INTERVAL = 1000;
const BLOCKS_TO_WAIT_FOR_BRIDGE = new BigNumber(14);

type BridgeTransactionInProgress = {
  startBalance: BigNumber;
  startEthBlock: BigNumber;
  expectedIncrease: BigNumber;
};

function useOnomyState() {
  const { blockNumber } = useContext<{ blockNumber: BigNumber }>(ChainContext);
  const blockNumRef = useRef(blockNumber);
  const onomyClient = useMemo(() => {
    return new OnomyClient(REACT_APP_ONOMY_REST_URL, REACT_APP_ONOMY_WS_URL);
  }, []);

  blockNumRef.current = blockNumber;
  const [address, setAddress, addressRef] = useStateRef('');
  const [amount, setAmount, amountRef] = useStateRef('0');
  const [bridgeTransactions, setBridgeTransactions] = useState<BridgeTransactionInProgress[]>([]);
  const [bridgedSupply, setBridgedSupply] = useState(new BigNumber(0));

  const bridgedSupplyFormatted = useMemo(() => {
    if (!bridgedSupply) return 0;
    const formated = format18(bridgedSupply);
    return formated.toNumber();
  }, [bridgedSupply]);

  const { address: keplrAddress, hasKeplr, connect: connectKeplr } = useKeplr(KEPLR_CONFIG);
  useEffect(() => {
    if (keplrAddress) setAddress(keplrAddress);
  }, [keplrAddress, setAddress]);

  const addPendingBridgeTransaction = useCallback((expectedIncrease: BigNumber) => {
    const transaction = {
      startBalance: new BigNumber(amountRef.current),
      startEthBlock: blockNumRef.current,
      expectedIncrease,
    };
    setBridgeTransactions(t => [...t, transaction]);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    // Clear any pending transactions that appear to have been processed
    const curBalance = new BigNumber(amount);
    const clearedTransactions = bridgeTransactions.filter(({ startBalance, expectedIncrease }) => {
      const expectedBalance = startBalance.plus(expectedIncrease);
      return curBalance.isGreaterThanOrEqualTo(expectedBalance);
    });
    setBridgeTransactions(pending => pending.filter(t => clearedTransactions.indexOf(t) === -1));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [amount]);

  const bridgeProgress = useMemo(() => {
    if (bridgeTransactions.length === 0) return null;
    const [{ startEthBlock }] = bridgeTransactions;
    const { startEthBlock: lastStartBlock } = bridgeTransactions.slice().pop()!;
    const expectedEndBlock = lastStartBlock.plus(BLOCKS_TO_WAIT_FOR_BRIDGE);
    const progressBlocks = blockNumber.minus(startEthBlock);
    const totalBlocks = expectedEndBlock.minus(startEthBlock);
    const progress = progressBlocks.dividedBy(totalBlocks).multipliedBy(100).toNumber();
    return Math.min(progress, 100);
  }, [blockNumber, bridgeTransactions]);

  useEffect(() => {
    async function updateBalance() {
      if (!addressRef.current) return;
      try {
        setAmount(await onomyClient.getAddressBalance(addressRef.current, DENOM));
      } catch (e) {
        // eslint-disable-next-line no-console
        console.warn('Error fetching Onomy balance', e);
      }
    }

    updateBalance();
    const interval = setInterval(updateBalance, POLLING_INTERVAL);
    return () => clearInterval(interval);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const updateBridgedSupply = useCallback(
    async () => setBridgedSupply(await onomyClient.getAnomSupply()),
    [onomyClient]
  );

  useEffect(() => {
    updateBridgedSupply();
    const interval = setInterval(() => updateBridgedSupply(), POLLING_INTERVAL);
    return () => clearInterval(interval);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return {
    address,
    amount,
    bridgedSupplyFormatted,
    bridgeProgress,
    hasKeplr,
    setAddress,
    addPendingBridgeTransaction,
    connectKeplr,
  };
}

export type OnomyState = ReturnType<typeof useOnomyState>;

const DEFAULT_STATE: OnomyState = {
  address: '',
  amount: '0',
  bridgedSupplyFormatted: 0,
  bridgeProgress: null,
  hasKeplr: false,
  connectKeplr: () => Promise.resolve(),
  setAddress: () => {},
  addPendingBridgeTransaction: () => {},
};

const OnomyContext = createContext(DEFAULT_STATE);

export function useOnomy() {
  const context = useContext(OnomyContext);
  context.connectKeplr();
  return context;
}

export function OnomyProvider({ children }: { children: JSX.Element | JSX.Element[] }) {
  const state = useOnomyState();

  return <OnomyContext.Provider value={state}>{children}</OnomyContext.Provider>;
}
