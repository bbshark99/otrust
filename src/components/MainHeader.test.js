import React from 'react';
import { cleanup, render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { BigNumber } from 'bignumber.js';

import { darkNew } from 'theme/theme';
import { ChainContext } from '../context/chain/ChainContext';
import MainHeader from './MainHeader';

describe('Given the MainHeader component and supplyNOM, currentETHPrice are of BigNumber type', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ThemeProvider theme={darkNew}>
          <ChainContext.Provider
            value={{ supplyNOM: new BigNumber(0), currentETHPrice: new BigNumber(10) }}
          >
            <MainHeader />
          </ChainContext.Provider>
        </ThemeProvider>
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

describe('Given the MainHeader component and supplyNOM, currentETHPrice are not of BigNumber type', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <ThemeProvider theme={darkNew}>
          <ChainContext.Provider value={{ supplyNOM: 0, currentETHPrice: 10 }}>
            <MainHeader />
          </ChainContext.Provider>
        </ThemeProvider>
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
