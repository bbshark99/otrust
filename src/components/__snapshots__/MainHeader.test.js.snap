// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the MainHeader component and supplyNOM, currentETHPrice are not of BigNumber type when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <header>
    <div
      class="sc-bdnxRM iNmwGW"
    >
      <header
        class="sc-iCoGMd jReHeF"
      >
        <div
          class="sc-fujyAs ipUnbt"
        >
          <img
            alt="Onomy"
            class="sc-pNWdM deYXhl"
            src="logo.svg"
          />
          <div
            class="sc-jrsJWt jghoxO"
          >
            <strong>
              Onomy
            </strong>
            <span>
              Bonding curve
            </span>
          </div>
        </div>
        <div
          class="sc-kEqXSa jsxILd"
        >
          <div
            class="sc-crzoAE fyXMGr"
          >
            <div
              class="sc-iqAclL gZIDje"
            >
              <strong>
                NOM / USDT
              </strong>
              <div
                class="sc-bqGGPW hYhdEA"
              >
                <strong>
                  $10.12
                </strong>
                <span
                  class="sc-ksluID kvKAf"
                  type="increase"
                >
                  24.2%
                </span>
              </div>
            </div>
            <div
              class="sc-iqAclL gZIDje"
            >
              <strong>
                NOM / ETH
              </strong>
              <div
                class="sc-bqGGPW hYhdEA"
              >
                <strong>
                  <span>
                    Loading
                  </span>
                </strong>
              </div>
            </div>
          </div>
          <div
            class="sc-dIsUp ixoOJO"
          >
            <div
              class="sc-iqAclL gZIDje"
            >
              <strong>
                NOM Issued
              </strong>
              <div
                class="sc-bqGGPW hYhdEA"
              >
                <strong>
                  <span>
                    Loading
                  </span>
                </strong>
                <span
                  class="sc-ksluID gwEcki"
                >
                  / 1E8
                </span>
              </div>
            </div>
          </div>
        </div>
      </header>
    </div>
  </header>
</DocumentFragment>
`;

exports[`Given the MainHeader component and supplyNOM, currentETHPrice are of BigNumber type when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <header>
    <div
      class="sc-bdnxRM iNmwGW"
    >
      <header
        class="sc-iCoGMd jReHeF"
      >
        <div
          class="sc-fujyAs ipUnbt"
        >
          <img
            alt="Onomy"
            class="sc-pNWdM deYXhl"
            src="logo.svg"
          />
          <div
            class="sc-jrsJWt jghoxO"
          >
            <strong>
              Onomy
            </strong>
            <span>
              Bonding curve
            </span>
          </div>
        </div>
        <div
          class="sc-kEqXSa jsxILd"
        >
          <div
            class="sc-crzoAE fyXMGr"
          >
            <div
              class="sc-iqAclL gZIDje"
            >
              <strong>
                NOM / USDT
              </strong>
              <div
                class="sc-bqGGPW hYhdEA"
              >
                <strong>
                  $10.12
                </strong>
                <span
                  class="sc-ksluID kvKAf"
                  type="increase"
                >
                  24.2%
                </span>
              </div>
            </div>
            <div
              class="sc-iqAclL gZIDje"
            >
              <strong>
                NOM / ETH
              </strong>
              <div
                class="sc-bqGGPW hYhdEA"
              >
                <strong>
                  <span>
                    0
                  </span>
                </strong>
              </div>
            </div>
          </div>
          <div
            class="sc-dIsUp ixoOJO"
          >
            <div
              class="sc-iqAclL gZIDje"
            >
              <strong>
                NOM Issued
              </strong>
              <div
                class="sc-bqGGPW hYhdEA"
              >
                <strong>
                  <span>
                    0
                  </span>
                </strong>
                <span
                  class="sc-ksluID gwEcki"
                >
                  / 1E8
                </span>
              </div>
            </div>
          </div>
        </div>
      </header>
    </div>
  </header>
</DocumentFragment>
`;
